libpod/healthcheck_linux.go

```diff
@@ -106,31 +106,43 @@ func (c *Container) removeTransientFiles(ctx context.Context, isStartup bool) er
 	// clean up as much as possible.
 	stopErrors := []error{}
 
-	// Stop the timer before the service to make sure the timer does not
-	// fire after the service is stopped.
-	timerChan := make(chan string)
-	timerFile := fmt.Sprintf("%s.timer", c.hcUnitName(isStartup))
-	if _, err := conn.StopUnitContext(ctx, timerFile, "ignore-dependencies", timerChan); err != nil {
-		if !strings.HasSuffix(err.Error(), ".timer not loaded.") {
-			stopErrors = append(stopErrors, fmt.Errorf("removing health-check timer %q: %w", timerFile, err))
+	// Remove both the bare unit name and the version with appended run count.
+	// Ensures backwards compat with older Podman where we would not have had a run counter.
+	unitNames := []string{c.hcUnitName(isStartup, false), c.hcUnitName(isStartup, true)}
+	for _, unitName := range unitNames {
+		// Stop the timer before the service to make sure the timer does not
+		// fire after the service is stopped.
+		timerChan := make(chan string)
+		timerFile := fmt.Sprintf("%s.timer", unitName)
+		if _, err := conn.StopUnitContext(ctx, timerFile, "ignore-dependencies", timerChan); err != nil {
+			if !strings.HasSuffix(err.Error(), ".timer not loaded.") {
+				stopErrors = append(stopErrors, fmt.Errorf("removing health-check timer %q: %w", timerFile, err))
+			}
+		} else if err := systemdOpSuccessful(timerChan); err != nil {
+			stopErrors = append(stopErrors, fmt.Errorf("stopping systemd health-check timer %q: %w", timerFile, err))
```

Comment:
This is ugly but I don't see a good way around it, also I really think if the stop is successful we know the first unit existed and should break out of the loop and not try the other name for no reason.

Reply:
Problem: By the time we clean up the healthcheck for a stopped container, the container has already stopped and the PID in the state is reset to 0 - so we don't know what unit to remove.

Reply:
That SGTM, definitely simpler.

Reply:
Reworked to do that and repushed